CODE_STYLE.md

```diff
@@ -0,0 +1,3293 @@
+# Prelude
+
+This guide is intended to become a reference for logstash developers, a central place where contributor can refer when in doubt about how to style their code. Nevertheless an style guide is also useful to make the code look similar, reducing complexities and especially when groups of contributors are growing.
+
+This style guide is based on the [guide][rails-style-guide] created by Bozhidar Batsov for [Rubocop][RuboCop], The Airbnb [guide][Airbnb], the Logstash team inputs, and others. There is also a rubocop file used to check this guide automatically.
+
+## Table of Contents
+
+* [Source Code Layout](#source-code-layout)
+* [Syntax](#syntax)
+* [Naming](#naming)
+* [Comments](#comments)
+  * [Comment Annotations](#comment-annotations)
+* [Classes](#classes--modules)
+* [Exceptions](#exceptions)
+* [Collections](#collections)
+* [Strings](#strings)
+* [Regular Expressions](#regular-expressions)
+* [Percent Literals](#percent-literals)
+* [Metaprogramming](#metaprogramming)
+
+## Source Code Layout
+
+  * Use `UTF-8` as the source file encoding. If you're using a ruby version older than 2.0, adding ```# encoding: utf-8``` on top of a ruby file will tell the compiler to use this encoding by default. If using a ruby VM older than 2.0, `UTF-8` is the default encoding.
+
+ * Use two **spaces** per indentation level (aka soft tabs). No hard tabs.
+
+  ```Ruby
+  # bad - four spaces
+  def some_method
+      do_something
+  end
+
+  # good
+  def some_method
+    do_something
+  end
+  ```
+
+ *  Don't use `;` to separate statements and expressions. As a corollary - use one expression per line.
+
+  ```Ruby
+  # bad
+  puts 'foobar'; # superfluous semicolon
+
+  puts 'foo'; puts 'bar' # two expressions on the same line
+
+  # good
+  puts 'foobar'
+
+  puts 'foo'
+  puts 'bar'
+
+  puts 'foo', 'bar' # this applies to puts in particular
+  ```
+
+  * Prefer a single-line format for class definitions with no body.
+
+  ```Ruby
+  # bad
+  class FooError < StandardError
+  end
+
+  # okish
+  class FooError < StandardError; end
+
+  # good
+  FooError = Class.new(StandardError)
```

Comment:
We may want to explicitly discourage using Class.new (and Module.new). See https://bugs.ruby-lang.org/issues/11119